### ISSUE TITLE
Is it reasonable to just average evaluated value?

### ISSUE THREAD

**Issue by wpstmxhs** [gcp/leela-zero#855]
As you know, there are some different approaches to utilize value network's evaluation.

For example, An open sourced computer go program RN, just averages value network evaluations and uses it, like Leela Zero. but another open sourced computer go program AQ, doesn't simply average, but averages in depth-weighted manner. It makes sense, because deeper evaluation of value network is more accurate than shallower one.

I want to discuss about this and listen everyone's opinions.

**Comment by jkiliani** [comment]
How does the UCT search done by AQ work? Also, what neural network architecture do they currently use? (2.1.1)

**Comment by wpstmxhs** [comment]
And I have an new idea. See the below picture below first.

This game my supervised 20-block Leela Zero bot lost by tygem 9D human, because bot played a severe blunder.

![board](https://user-images.githubusercontent.com/28377280/36072840-43932df2-0f6a-11e8-9699-32cc282a089b.png) (Result: W+Resign)

And the following is problematic situation which my bot caused a bug and the next one is its evaluation of candidate moves.

![board2](https://user-images.githubusercontent.com/28377280/36072848-78082cfe-0f6a-11e8-83c5-b6f55dd9e132.png) (Black to play)

```
 D19 ->    3139 (V: 89.48% / MV: 88.05%) (N:  2.99%) PV: D19 H1 J2 G1 J7 B19 G15 L14 F15 E16 N15 P12 Q12 P5 O6 C9 B9 A15 A14 M16
  S2 ->    2712 (V: 89.79% / MV: 89.79%) (N:  0.13%) PV: S2 S3 G1 J2 J3 H1 J4 J7 G2 H2 G15
  H1 ->    2665 (V: 89.54% / MV: 89.13%) (N:  9.07%) PV: H1 P12 Q12 L14 G15 N15 N16 P5 O6 F15 G13 C9 B9 A15 A14 B19
 G15 ->    1369 (V: 88.33% / MV: 86.84%) (N: 18.28%) PV: G15 P12 Q12 H1 J2 G1 J7 C9 B9 P5 O6 N15 N16 L14 S19 R19 T17
  A6 ->    1226 (V: 88.11% / MV: 88.09%) (N: 19.87%) PV: A6 A4 D19 B19 G1 J2 J7 P5
 N15 ->    1000 (V: 89.15% / MV: 87.37%) (N:  2.85%) PV: N15 H1 J2 G1 J7 P12 Q12 P5 O6 C9 B9 L14 G15 F15 G13 A15
 L14 ->     853 (V: 88.20% / MV: 86.09%) (N: 11.23%) PV: L14 P12 Q12 H1 J2 G1 J7 P5 O6 C9 B9 G15 D19
 P12 ->     564 (V: 88.23% / MV: 86.07%) (N:  7.00%) PV: P12 H1 J2 G1 J7 P5 O6 G15 L14 C9 B9 A15 A14 B19
  J7 ->     365 (V: 89.30% / MV: 87.73%) (N:  1.27%) PV: J7 H1 J2 G1 L14 G15 P12 P5 O6 C9 B9 A15 A14
  J2 ->     250 (V: 89.25% / MV: 86.91%) (N:  0.45%) PV: J2 G1 H1 P12 Q12 P5 O6 C9 B9 L14 G15 N15
  G1 ->     171 (V: 77.46% / MV:  0.30%) (N: 20.59%) PV: G1 J2 J7 H1 J4 P12 Q12 P5 O6
 A15 ->      66 (V: 88.47% / MV: 84.59%) (N:  0.63%) PV: A15 H1 J2 G1 J7 N15 N16 P12
  P5 ->      61 (V: 87.37% / MV: 81.93%) (N:  1.30%) PV: P5 N6 O6 P12 Q12 H1 J2 G1 J7 L14 G15 C9
  C9 ->      41 (V: 87.60% / MV: 84.71%) (N:  0.70%) PV: C9 H1 J2 G1 J7 P5 L14 G15 P12
 L17 ->      20 (V: 84.00% / MV: 80.54%) (N:  1.11%) PV: L17 N15 N16 L14 G15 L15 M16
  J3 ->       9 (V: 82.74% / MV: 76.15%) (N:  0.40%) PV: J3 H1 J4
 F15 ->       3 (V: 74.47% / MV: 73.73%) (N:  0.48%) PV: F15 G15 L14
14514 visits, 1707264 nodes, 14514 playouts, 406 n/s
```

There are many alternative moves but bot should not play G1.
You can see G1's winning rate is saying it's winning, but MV(I added this feature, which means minimal my value on opponent candidate moves) is 0.30%.
I think G1's winning rate should be 0.30% or somewhat closer to 0.30%.

So, how about to use principal variation value instead of averaged-all value?

**Comment by RavnaBergsndot** [comment]
Deeper evaluation ignores possible situations when the neural network chose the wrong path to go deeper. So it's not necessarily more accurate.

**Comment by wpstmxhs** [comment]
@jkiliani AQ's MCTS works by just combining random rollout winning rate and value network evaluation. I'm not so familiar to AQ's architecture, but AQ's network seems based on old AlphaGo's network architecture. It gets 48-channeled input.

**Comment by wpstmxhs** [comment]
@RavnaBergsndot Yeah, I agree that. but I think it's good if we have some propagation method of deeper evaluation. I don't think current approach is the best.

**Comment by RavnaBergsndot** [comment]
The problem of using the principal valuation was summarized in the AGZ paper.

"AlphaZero evaluates positions using non-linear function approximation based on a deep neural network, rather than the linear function approximation used in typical chess programs. This provides a much more powerful representation, but may also introduce spurious approxi- mation errors. MCTS averages over these approximation errors, which therefore tend to cancel out when evaluating a large subtree. In contrast, alpha-beta search computes an explicit mini- max, which propagates the biggest approximation errors to the root of the subtree." (page 12)

**Comment by wpstmxhs** [comment]
@RavnaBergsndot So AlphaGo team says current approach is best and many self training will solve that well?

Anyway, I saw AQ outperforms against supervised 20-block Leela Zero. Maybe pure Zero network will do better after many self training?

**Comment by RavnaBergsndot** [comment]
I don't know. This is their only comment about MCTS vs Alpha-Beta in that paper. They didn't show the examples of "approximation errors" and how they damage the playing strength.

AQ outperforms agains 20-block Leela Zero probably because of its added rollout module. Even Alphago Master has the rollout module. This tells us, at least before 5000 ELO, rollouts make the engine stronger.

I heard there's a bot which mixed the LZ neural network and the RN rollouts. It reached Fox 9d when LZ was still at 6d or 7d.

**Comment by wpstmxhs** [comment]
I see. I also did that before. I mixed Leela Zero into Ray. it won against many tygem 9D players.

**Comment by jkiliani** [comment]
I think the supervised 20 block LZ is a lot weaker than reinforcement learning will achieve on 20x256, perhaps not even stronger than 10x128 will end up. Even for 6x128 the RL nets are already a lot better than SL, and for larger networks this disparity will be even larger, since the quality of available training data is another limiting factor in how strong a supervised net for LZ can be.

**Comment by wpstmxhs** [comment]
@jkiliani Yes I think so too. Generally speaking, I think reinforcement learning performs better than supervised learning on the same sized network finally.

**Comment by roy7** [comment]
@wpstmxhs Do you have a fork with LZ with a branch with your MV changes I can try out? I've been thinking about trying to learn this area of the code better so I can try some ideas myself. Generally I wanted to try adjusting move selection away from "most visits" to something else (like using the confidence intervals on the win rates, and going with the highest lower bound in a "minimize regret" sort of approach).

I've never looked at AQ's source code, but not it's by far the strongest open source Go bot out there validated by CGOS rank. I don't know how much of that strength is from the neural network training vs how it does search or other tricks.

**Comment by roy7** [comment]
@wpstmxhs On the bots I run I use --log with a filename based on the game ID from the server, so I have logs of every game the bot has played. That'd let you go back and see the exact search logs on that G1 move instead of need to do the search fresh after the fact which will have different randomization/etc results.

**Comment by wpstmxhs** [comment]
@roy7 Oh, My `MV(My Minimal Value on opponent side)` change was just adding debug message. not changing any move reading logic. I'm avoiding that blunders by checking MV value and do another move manually by human. And I have idea to change reading code anyhow, but I have no idea about how to do it. That's why I want to discuss about these problems with you all.

```cpp
		double min_winrate = 5.0; // prints 500% if no visits
		for (const auto &child_node : node->get_children()) {
			if (!child_node->get_visits()) continue;
			double cur_score = child_node->get_eval(color);
			if (cur_score < min_winrate) min_winrate = cur_score;
		}

        myprintf("%4s -> %7d (V: %5.2f%% / MV: %5.2f%%) (N: %5.2f%%) PV: ",
            tmp.c_str(),
            node->get_visits(),
            node->get_eval(color)*100.0f,
			min_winrate * 100.0f,
            node->get_score() * 100.0f);
```

**Comment by roy7** [comment]
Thanks! I'll add it in and start checking it out. I'm slowly trying to learn more about how the search works. I did come across a paper called [Bandit Algorithms for Tree Search](https://arxiv.org/ftp/arxiv/papers/1408/1408.2028.pdf) which introduces an approach called Bandit Algorithm for Smooth Trees (BAST) which seems to have less regret than Upper Confidence Bounds applied to Trees (UCT). I'm also curious about other approaches to tree search not based on Upper Confidence Bounds (UCB) at all.

**Comment by gcp** [comment]
>And the following is problematic situation which my bot caused a bug

What do you mean, it played G1 there? Do you have randomization enabled? Or is the search you were showing a separate, deeper instance?

**Comment by gcp** [comment]
>just averages value network evaluations and uses it, like Leela Zero. but another open sourced computer go program AQ, doesn't simply average, but averages in depth-weighted manner

This is an extremely misleading statement. The value in the node is the average of all playouts that go through it. More playouts towards a child means that it is explored deeper. So, the value in a current node **already is** weighted by the depth the subtrees have been explored.

**Comment by wpstmxhs** [comment]
> @gcp What do you mean, it played G1 there? Do you have randomization enabled? Or is the search you were showing a separate, deeper instance?

Yes, It played it on real game with supervised 20-block net trained by @zediir . I wanted to test with full strength of it, of course I didn't enable the randomization feature.

**Comment by wpstmxhs** [comment]
>The value in the node is the average of all playouts that go through it. More playouts towards a child means that it is explored deeper.

I see. but when Leela Zero could not find the best variation and is trying to find alternatives more and more, I think extra weight may be useful.

**Comment by gcp** [comment]
So is the output you showed from the actual search? Or is that a test afterwards

If it played G1 there with randomization disabled and that is the actual search output this is a massive bug.

**Comment by gcp** [comment]
I'm asking because G1 has a huge policy prior, so if it moved quickly in the game or if it did a narrow search then it is likely to have been played. But the search you showed shouldn't ever play it.

**Comment by wpstmxhs** [comment]
@gcp Yes. on actual game, it played G1 with not so small playouts. i think it depends on random seed initialized on program startup.

**Comment by wpstmxhs** [comment]
For debugging purpose, I upload the actual game sgf file of that game.

```
(;GM[1]FF[4]AP[Zenith:7.0]SZ[19]HA[0]KM[6.5]CA[UTF-8]PB[wpstmxhs]BR[]
PW[bnyyzy]WR[]ZT[5]DT[]RE[W+Resign];B[pd];W[dp];B[qp];W[dc];B[fq];W[oq];
B[cn];W[dn];B[dm];W[en];B[co];W[cp];B[em];W[fm];B[fl];W[fn];B[gl];W[hq];
B[gp];W[hp];B[go];W[cm];B[cl];W[bm];B[bl];W[ho];B[gn];W[gm];B[hn];W[hm];
B[in];W[kp];B[bn];W[el];B[dl];W[do];B[kn];W[bo];B[am];W[po];B[dr];W[cr];
B[dq];W[hl];B[jl];W[jk];B[kk];W[jj];B[ek];W[kj];B[il];W[hk];B[ik];W[ij];
B[hj];W[gk];B[gj];W[fj];B[fi];W[cq];B[lj];W[li];B[lk];W[ej];B[ei];W[dj];
B[di];W[ao];B[jh];W[cj];B[ci];W[ii];B[hh];W[ih];B[ig];W[jg];B[hg];W[fk];
B[bj];W[kh];B[jf];W[kg];B[mp];W[lo];B[no];W[pp];B[nm];W[ln];B[ml];W[lm];
B[kl];W[mq];B[lq];W[np];B[mr];W[nq];B[qm];W[qo];B[aq];W[ar];B[jp];W[kq];
B[jq];W[kr];B[jr];W[jo];B[io];W[jn];B[ks];W[ls];B[js];W[lr];B[ni];W[ng];
B[og];W[of];B[pg];W[nh];B[mi];W[mh];B[oi];W[ce];B[qe];W[bh];B[bi];W[pm];
B[pl];W[ql];B[rm];W[om];B[qk];W[ol];B[pk];W[ob];B[oc];W[nc];B[nd];W[pc];
B[od];W[qb];B[mc];W[nb];B[qc];W[pb];B[rc];W[mb];B[db];W[cb];B[eb];W[ec];
B[fc];W[jc];B[bb];W[cc];B[ca];W[fd];B[gc];W[ba];B[aa];W[da];B[ea];W[bc];
B[bf];W[cf];B[bg];W[gd];B[hc];W[ie];B[he];W[hd];B[ic];W[jd];B[jb];W[kb];
B[id];W[je];B[ia];W[fr];B[gr];W[gq];B[hr];W[eq];B[fp];W[er];B[ff];W[ac];
B[lc];W[lb];B[ro];W[rp];B[rn];W[qq];B[le];W[mf];B[on];W[pn];B[cg];W[ee];
B[ef];W[df];B[dg];W[rb];B[sb];W[pf];B[qf];W[ga];B[gb];W[nl];B[mm];W[oj];
B[pj];W[nj];B[mj];W[ka];B[mo];W[lp];B[ha];W[hi];B[gi];W[ib];B[fa];W[ab];
B[be];W[bd];B[ja];W[hf];B[gf];W[if];B[lf];W[lg];B[km];W[ko];B[so];W[el];
B[fl];W[fs];B[ep];W[ds];B[ne];W[nf];B[sp];W[sq];B[bp];W[bq];B[gs];W[ir])
```

**Comment by wpstmxhs** [comment]
I just repeated three times. On third try, the latest Leela Zero (of next branch) played G1 with 5000 playouts.

Here's console output:

```

Clamping threads to maximum = 16
RNG seed: 7400904865014522428
Detecting residual layers...v1...256 channels...20 blocks.
Initializing OpenCL.
Detected 1 OpenCL platforms.
Platform version: OpenCL 1.2 CUDA 9.1.75
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     0
Device name:   GeForce GTX 1070
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 388.13
Device speed:  1708 MHz
Device cores:  15 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1070
with OpenCL 1.2 capability.
Loaded existing SGEMM tuning.
Wavefront/Warp size: 32
Max workgroup size: 1024
Max workgroup dimensions: 1024 1024 64 
BLAS Core: Barcelona
undo
= 

genmove b
Thinking at most 50.0 seconds...
NN eval=0.988653
Playouts: 672, Win: 86.01%, PV: G1 P12 Q12 P5 G15 O6 N6 L14 N15 C9
Playouts: 1662, Win: 89.89%, PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
Playouts: 2622, Win: 88.94%, PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
Playouts: 3614, Win: 89.09%, PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
Playouts: 4620, Win: 89.52%, PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
Stopping early.

  G1 ->     783 (V: 89.39%) (N: 20.59%) PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
 D19 ->     622 (V: 92.56%) (N:  2.99%) PV: D19 B19 G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15
  T2 ->     580 (V: 93.10%) (N:  0.05%) PV: T2 S3 G1 P12 Q12 L14 G15 N15 N16 L15 M16 C9
  H1 ->     526 (V: 90.61%) (N:  9.07%) PV: H1 P12 Q12 P5 O6 L14 N15 G15 C9 A15 A14 M16
 G15 ->     428 (V: 88.56%) (N: 18.28%) PV: G15 H1 J2 G1 J7 P5 O6 P12 Q12 N15 N16 L14 C9
 L14 ->     417 (V: 89.77%) (N: 11.23%) PV: L14 L17 N15 M16 N16 P12 Q12 H1 J2 G1 J7 P5 O6 C9 B9 G15
  A6 ->     337 (V: 87.72%) (N: 19.87%) PV: A6 A4 G1 P12 Q12 N15 N16 L14 G15 L15 M16 P5
 N15 ->     321 (V: 90.34%) (N:  2.85%) PV: N15 P12 Q12 H1 J2 G1 J7 P5 O6 G15 L14 C9
  S3 ->     319 (V: 91.70%) (N:  0.11%) PV: S3 S2 G1 P12 Q12 L14 N15 P5 O6 C9 B9 G15 A15 B19
 P12 ->     305 (V: 89.62%) (N:  7.00%) PV: P12 H1 J2 G1 J7 P5 G15 O6 N6 L14 N15
  P5 ->      62 (V: 89.72%) (N:  1.30%) PV: P5 N6 O6 P12 Q12 L14 N15 H1 J2 G1
  C9 ->      46 (V: 89.90%) (N:  0.70%) PV: C9 P12 Q12 P5 O6 H1
  J7 ->      29 (V: 88.04%) (N:  1.27%) PV: J7 H1 J2 G1 G15 P5 O6
  J2 ->      22 (V: 88.76%) (N:  0.45%) PV: J2 P12 Q12 N15 N16 L14
 Q15 ->      11 (V: 84.29%) (N:  0.04%) PV: Q15 P12 Q12 H1 J2
  J3 ->      10 (V: 82.04%) (N:  0.40%) PV: J3 H1 J4 G1 L14
 A15 ->       9 (V: 87.18%) (N:  0.63%) PV: A15 H1 J2 G1
 L17 ->       9 (V: 82.09%) (N:  1.11%) PV: L17 L14 N15 P12
 B19 ->       7 (V: 81.52%) (N:  0.07%) PV: B19 D19 C19 B19 L14
 S19 ->       7 (V: 79.43%) (N:  0.08%) PV: S19 H1 J2
  J1 ->       1 (V: 79.26%) (N:  0.04%) PV: J1 
  S2 ->       1 (V: 78.53%) (N:  0.13%) PV: S2 
  J4 ->       1 (V: 78.00%) (N:  0.09%) PV: J4 
 F15 ->       1 (V: 75.97%) (N:  0.48%) PV: F15 
 L15 ->       1 (V: 71.35%) (N:  0.02%) PV: L15 
 T17 ->       1 (V: 68.00%) (N:  0.06%) PV: T17 
 D16 ->       1 (V: 67.55%) (N:  0.07%) PV: D16 
4857 visits, 578472 nodes, 4857 playouts, 364 n/s

= G1

```

**Comment by odeint** [comment]
> -t 9999999

Why are you trying to have 10 million threads? Isn't LZ strongest with 1 or 2 threads? At least in the output, it seems that the number of threads cannot go beyond 16 (is that your number of CPUs / SMT cores. Still, I think there was something negative about using many threads compared to the same number of playouts on a single one.

**Comment by wpstmxhs** [comment]
@odeint Yes. it's ok if I use `-t 16` on my system. but `-t <very big number>` automatically sets thread limit that the system supports.

And the big network like supervised 20-block network, it's efficient to use more threads. because one network evaluation takes long time, simultaneous evaluation should be efficient. You can see playout speed is only 364 n/s even though I use all the threads on my system.

**Comment by gcp** [comment]
Thanks, this is fine, G1 is the topmost move so it's OK to play it. No bug.

>And the big network like supervised 20-block network, it's efficient to use more threads. because one network evaluation takes long time, simultaneous evaluation should be efficient

Most people measured a slowdown from more threads. You are contending with the GPU driver for CPU usage, and only a few threads are needed to fully load the GPU.

**Comment by roy7** [comment]
@gcp The interesting thing about the blunder G1 move is that, although the win rate was too high and search didn't realize it was bad, his custom MV stat did show it as a far worse move than the other choices (opponent had candidate move after G1 to reduce our win rate to 0.3%).

**Comment by WhiteHalmos** [comment]
Hmm... are we propagating wrong? Since if that candidate move does exist, the winrate for the other color should be really high (99.7%) which should result in a very low value for G1 on every visit of that child node?

**Comment by roy7** [comment]
Well remember G1 is an average, so most of the searches must go down paths that black still wins the game. Only G1 J2 ends up losing if other G1 XX moves are being considered. Maybe in this case it never saw G1 J2, I'm not sure. When he went back and did the search and it came up with

> 171 (V: 77.46% / MV:  0.30%) (N: 20.59%) PV: G1 J2 J7 H1 J4 P12 Q12 P5 O6

It did drop win rate from ~88% to 77%. But observation is that since the opponent has a 0.30% response move he felt it should be a lot lower. The 77% is probably made up of lots of G1 P12 search (as you can see on the 3rd re-do when he could recreate it making that move) and a few searches of G1 J2, so it averages down from 88% to 77%.

In his most recent re-creation we see:

>   G1 ->     783 (V: 89.39%) (N: 20.59%) PV: G1 P12 Q12 N15 N16 L14 G15 P5 O6 C9 B9 A15 A14 B19
>  D19 ->     622 (V: 92.56%) (N:  2.99%) PV: D19 B19 G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15

Since it's 3% lower than D19 I'm guessing it found G1 J2 at the end of search right before actually making a move ,and before more search could change us to D19. It's a shame this didn't include the MV printout. I'd love to see what the MV % is in a case where G1 is the winning move and still thinks it is 89%.

**Comment by WhiteHalmos** [comment]
@wpstmxhs Would you mind creating an sgf that can reproduce this with a simple command like `echo 'loadsgf <name>.sgf 125\ngenmove b\nquit' | ./leelaz etc etc` so it's somewhat more reproducible?

**Comment by roy7** [comment]
@WhiteHalmos You'd need his weights file as well (custom 20 block supervised learning weights) but it'd definitely be useful for testing if we could easily reproduce an end game blunder like this where the immediate response move wins the game for the opponent.

I know minmax is inferior for our purposes with fuzzy board evaluations, and errors compounding, etc. But I can't help but think of it this way...

If we move to G1 and the opponent is searching for a response move, the prior N% for P12 must be high. That only gives him a 11% chance to win (since we're at 89%) but other moves must be similar or worse. In the blunder situation, G1 J2 was found late so the win rate for opponent starts to rise towards 99.7% the more search we put into G1.

If opponent happened to search J2 early by random chance, then almost all future search would be going into J2 since the opponent win rate is 99.7% vs 11% for other moves like P12. Thus the win rate for black here would lock in near optimal early and stay there. And the prior step, when we are looking at moves, we should see G1 J2 as like a 0.3% win rate right away if G1 J2 happened to be the first move we searched.

It's all about the timing of when we find J2 vs P12. But my thought is this... since we know once G1 J2 is found (99% win) that G1 P12 is long gone (11% win) and all future search will go into G1 J2 if more search goes into G1, why would we ever want to actually move to G1 P12? 

@wpstmxhs uses this approach to manually avoid blunders when relaying for Leela, but there's surely some sound theory by which we could detect a blunder and make an immediate change in the move we want to choose regardless of the relative # of visits between the moves. This might be applicable to the ladder situation as well, but it's nice we have an example blunder situation where a single move can cost you the game (which Leela v11 sometimes falls into as well when she loses in the end game).

**Comment by wctgit** [comment]
>It's all about the timing of when we find J2 vs P12. But my thought is this... since we know once G1 J2 is found (99% win) that G1 P12 is long gone (11% win) and all future search will go into G1 J2 if more search goes into G1, why would we ever want to actually move to G1 P12?

This is a good summary of the dilemma. I don't have an immediate answer, but it definitely stirs up some thoughts. One thought that I think I should mention right away is the concept of the [Sunk Cost Fallacy](https://en.wikipedia.org/wiki/Sunk_cost#Loss_aversion_and_the_sunk_cost_fallacy).

Essentially, it seems to me, that there are (sometimes unstated) underlying assumptions being made in different kinds of tree searches. In the minimax case, the assumption is you want to minimize the **maximum** loss of the game value. In the MCTS case, it seems to me that the underlying assumption is that you want to minimize the **expected** loss of the game value. And this is where I think the Sunk Cost Fallacy might apply.

To estimate an 'expectation' (or 'expected value' of some variable), you generally have to take many multiple samples, sum them up and average them. The more samples you use, generally the more accurate your estimate. Therefore, the implicit policy is: Don't throw away perfectly good data; if you want a better estimate, then just collect more samples. In other words, "We've already spent 5000 playouts on G1; if you want a more accurate assessment of G1, don't throw out those playouts; instead, collect more, but from the more probable G1, J2, rather than G1, P12."

In contrast, to find a maximum or minimum, you don't need to remember all the other numbers, you just need to know if the next number is bigger or smaller. Once you've found a new max or min, the old one can just be forgotten/thrown out. In other words, "Who care's how much effort we've sunk into G1, P12? G1, J2 is a bigger loss, and that's all that matters. Let's abandon it now rather than waste any more time on it."

The 'sunk cost' in this analogy is the time / computational resources spent on tree search. And it seems to be a fallacy because the actual 'utility' of the situation should be more about actually winning the game than about penny-pinching computation resources.

But on the other hand, computation resources generally == time, and in most Go games, time is an essential factor for **actually winning** the game or losing by timeout.

Therefore, my hunch is that the theoretical foundation you're looking for is in [Decision Theory](https://en.wikipedia.org/wiki/Decision_theory), and specifically the concept of [Expected Utility](https://en.wikipedia.org/wiki/Expected_utility_hypothesis).

By looking at the dilemma through the lens of 'utility', you can see that minimax and MCTS simply have different fundamental assumptions about how to measure utility. But given the utility assumptions of minimax, then minimax gives the right solution. However, in the current situation (this project, its goals, the game of Go, limitations on resources (some of them self-imposed, like -p 1600), personal intuitions about what 'should' be played, etc.), the 'correct' utility function is not as cut-and-dried as the minimax criterion. For example, we know that minimax searching doesn't work so well with the game of Go because of practical considerations of Go's branchiness and limited hardware speeds (not to mention the algorithmic complexity remaining essentially exponential and therefore ultimately unsustainable).

So, in a sense, the 'sunk costs' of the MCTS expectation calculation**do** actually represent something that should be incorporated into the 'correct' utility function, namely the costs of computation of playouts. On the one hand, MCTS makes faster-than-minimax evaluations possible; on the other hand, when playouts are limited, you may simply not have enough available to get the most accurate expectation calculation.

So, depending on our personal utility functions, we might say, "Well, just increase the number of playouts then!" But then we'll run into people who have different assessments of utility, and want the strongest Go bot for time-limited games, or perhaps blitz games, and they'll say, "No! We need to tweak the code to avoid this sunk cost fallacy!"

The point is, there are different ways to solve this meta-decision-optimization problem (i.e. "What method should we use to fix this bot's decision-optimization problem?"), but the 'correct' way to do so is to think about it in terms of Decision Theory, and our 'expected utility' of choosing this tweak or that tweak (or no tweaks). For example: 

How important is it (i.e. "What is the utility for us ...") that we arbitrarily limit the bot to a fixed number of playouts? How about when there are time controls? How important is preventing an occasional embarrassing "blunder" (egg on this project's face) versus keeping the principles 'pure' (e.g. no special human knowledge cooked in)? How much time do we want to spend on fixing this problem now, rather than later? How much 'utility' **is** there in the bot's winning a game of Go? From our perspective ("It's just a bot." "But it's OUR bot!!!")? From the bot's perspective ("It's just a job." "But it's literally the entire purpose to my existence!!! If I lose, I'm literally a failure!")? How much utility **should** winning the game have? From a theoretical perspective (what's the ideal/theoretically-optimal value)? From a practical perspective (what makes the bot actually win more games)? (I could go on. I hope you get the gist.)

So, while I don't have any concrete answers ("After all that verbiage?! Come on!"), I think it might be a good time to take a step back, examine the underlying assumptions of the current technique (using visit count as a proxy for likelihood and/or risk; holding on to existing playouts because in theory that's how you get better approximations of expected values of variables; etc.), examine our own (i.e. this project's) values and what kinds of solutions we would find of highest 'utility', look for alternative methods perhaps, brainstorm some workarounds or tweaks perhaps, decide how important this issue is to resolve, etc.

Of course, we're already basically doing that kind of thing in this very thread. My point is to do it consciously, challenging our implicit assumptions and making them more explicit, tying it into Decision Theory (for the bot and for us), etc.

I hope this was more useful than not. :-)

**Comment by wpstmxhs** [comment]
@gcp 

> You are contending with the GPU driver for CPU usage, and only a few threads are needed to fully load the GPU.

I just thought that it's near optimal if I use more threads as many as possible. because of thread scheduling of OS, each thread after once network evaluation doesn't get a chance to resume right soon. But, due to locking/synchronizing overhead, It may slow down a little. I think network size matters in this case. To be honestly, I didn't do any scientific test to measure performance on each thread settings.

**Comment by wpstmxhs** [comment]
@roy7 

> It's a shame this didn't include the MV printout. I'd love to see what the MV % is in a case where G1 is the winning move and still thinks it is 89%.

I remade that problem with MV info printed. See the below:

```
genmove b
Thinking at most 50.0 seconds...
NN eval=0.987617
Playouts: 626, Win: 85.49%, PV: G1 P12 Q12 P5 O6 L14 G15 N15
Playouts: 1556, Win: 89.42%, PV: G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15
Playouts: 2496, Win: 89.23%, PV: G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15 A14
Playouts: 3449, Win: 89.43%, PV: G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15 A14
Playouts: 4511, Win: 89.75%, PV: G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15 A14
Stopping early.

  G1 ->     756 (V: 90.16% / MV:  5.32%) (N: 13.92%) PV: G1 P12 Q12 P5 O6 L14 N15 G15 C9 A15 A14
 D19 ->     673 (V: 93.06% / MV: 90.88%) (N:  4.73%) PV: D19 B19 G1 P12 Q12 L14 N15 G15 A15 C9 B9 P5
  A6 ->     497 (V: 91.15% / MV: 91.16%) (N: 13.04%) PV: A6 A4 G1 P12 Q12 P5 O6 C9 B9 N15 N16 L14
 N15 ->     474 (V: 90.70% / MV: 88.97%) (N:  7.68%) PV: N15 H1 J2 G1 J7 P12 Q12 P5 O6 L14 G15
 L14 ->     445 (V: 89.26% / MV: 85.73%) (N: 15.68%) PV: L14 N15 N16 H1 J2 G1 J7 G15 P12 P5 O6 C9
 G15 ->     422 (V: 89.32% / MV: 86.64%) (N: 16.32%) PV: G15 H1 J2 G1 J7 N15 N16 L14 P12 L15 M16 P5
  S2 ->     369 (V: 92.28% / MV: 92.30%) (N:  0.11%) PV: S2 S3 G1 P12 Q12 L14 N15 G15 C9 A15 A14
 P12 ->     324 (V: 90.31% / MV: 88.60%) (N:  8.08%) PV: P12 H1 J2 G1 J7 P5 O6 L14 G15 C9 B9 N15 N16
  H1 ->     304 (V: 90.96% / MV: 89.44%) (N:  3.27%) PV: H1 P12 Q12 P5 O6 L14 N15 G15 C9 M16
  J7 ->     254 (V: 90.73% / MV: 89.37%) (N:  3.34%) PV: J7 H1 J2 G1 G15 N15 N16 L14 P12 L15 M16 P5
  S3 ->     194 (V: 92.17% / MV: 92.21%) (N:  0.09%) PV: S3 S2 G1 P5 G15 L14 N15 P12 Q12
  P5 ->      89 (V: 86.72% / MV: 85.81%) (N:  7.10%) PV: P5 N6 O6 H1 J2 G1 J7 P12 Q12 G15 L14 C9
  C9 ->      37 (V: 89.17% / MV: 88.28%) (N:  1.37%) PV: C9 H1 J2 G1 J7 P12 Q12 P5
  J2 ->      22 (V: 89.99% / MV: 88.93%) (N:  0.08%) PV: J2 G1 H1 P12 Q12 L14
 Q15 ->      16 (V: 88.00% / MV: 86.54%) (N:  0.91%) PV: Q15 P12 Q12 H1 J2
  J1 ->      13 (V: 87.70% / MV: 87.03%) (N:  0.04%) PV: J1 G1 H1 P5 O6 P12
 L17 ->      10 (V: 86.39% / MV: 82.51%) (N:  0.93%) PV: L17 N15 N16 L14
  J3 ->       9 (V: 83.29% / MV: 72.11%) (N:  0.30%) PV: J3 P12 Q12 N15 N16
 B19 ->       8 (V: 86.78% / MV: 86.07%) (N:  0.04%) PV: B19 D19 G1 P12 Q12
  J4 ->       5 (V: 82.77% / MV: 76.79%) (N:  0.05%) PV: J4 H1 J3
 F15 ->       5 (V: 78.27% / MV: 76.05%) (N:  0.86%) PV: F15 G15 G1
 L15 ->       1 (V: 82.80% / MV: N/A) (N:  0.02%) PV: L15 
 D16 ->       1 (V: 81.22% / MV: N/A) (N:  0.05%) PV: D16 
 M16 ->       1 (V: 75.55% / MV: N/A) (N:  0.03%) PV: M16 
 A15 ->       1 (V: 72.91% / MV: N/A) (N:  0.55%) PV: A15 
 S19 ->       1 (V: 72.19% / MV: N/A) (N:  0.36%) PV: S19 
 L11 ->       1 (V: 65.85% / MV: N/A) (N:  0.03%) PV: L11 
4932 visits, 588152 nodes, 4932 playouts, 359 n/s

= G1


```

**Comment by roy7** [comment]
Perfect, thank you. So the MV "blunder warning" does work in that case. I see you've adopted it into a search experiment at https://github.com/gcp/leela-zero/issues/868. Very exciting. :)
